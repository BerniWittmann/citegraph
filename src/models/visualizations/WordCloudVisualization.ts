import Visualization, { VisualizationBaseFields } from './Visualization'

export interface WordCloudVisualizationParameters {
  pruneLeastCitedPercentage: number
}

export interface WordCloudDataPoint {
  word: string
  weight: number
}
export interface WordCloudData {
  data: Array<WordCloudDataPoint>
}

export interface WordCloudVisualizationFields {
  data?: WordCloudData
}

export default class WordCloudVisualization extends Visualization implements WordCloudVisualizationFields {
  static key: string = 'wordcloud'
  key: string = 'wordcloud'
  static title: string = 'visualizations.information.word_cloud.title'
  static description: string = 'visualizations.information.word_cloud.description'
  static longDescription: string = 'visualizations.information.word_cloud.long_description'
  static imageUrl: string = 'visualization_images/word_cloud.jpg'
  parameters: WordCloudVisualizationParameters
  data?: WordCloudData

  constructor ({ id, name, progress, data }: VisualizationBaseFields & WordCloudVisualizationFields) {
    super({ id, name, progress })
    this.parameters = {
      pruneLeastCitedPercentage: 5
    }
    this.data = data
  }
}
