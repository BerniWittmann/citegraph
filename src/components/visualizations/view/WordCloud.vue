<template>
  <v-container>
    <v-row justify="center">
      <v-col cols="auto">
        <v-card class="pa-5">
          <div ref="vis"></div>
        </v-card>
      </v-col>
    </v-row>
  </v-container>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'
import WordCloudVisualization, {
  WordCloudDataPoint
} from '@/models/visualizations/WordCloudVisualization'
import * as d3 from 'd3'
import cloud from 'd3-cloud'

@Component
export default class WordCloud extends Vue {
  @Prop() readonly visualization!: WordCloudVisualization

  width = 500
  height = 500
  layout: any

  get diagramData (): Array<WordCloudDataPoint> {
    return this.visualization.data!.data
  }

  draw (words: any): void {
    const fill = d3.scaleOrdinal(d3.schemeCategory10)
    const svg = d3.select(this.$refs.vis as HTMLElement).append('svg')
      .attr('width', this.layout.size()[0])
      .attr('height', this.layout.size()[1])

    const g = svg.append('g')
      .attr('transform', 'translate(' + this.layout.size()[0] / 2 + ',' + this.layout.size()[1] / 2 + ')')

    g.selectAll('text')
      .data(words)
      .enter().append('text')
      .style('font-size', (d: any) => d.size + 'px')
      .style('font-family', 'Impact')
      .style('fill', (d: any, i: number) => fill(i.toString()))
      .attr('text-anchor', 'middle')
      .attr('transform', (d: any) => 'translate(' + [d.x, d.y] + ')rotate(' + d.rotate + ')')
      .text((d: any) => d.text)
  }

  mounted () {
    this.layout = cloud()
      .size([this.width, this.height])
      .words(this.diagramData.map(function (d: WordCloudDataPoint) {
        return { text: d.word, size: d.weight }
      }))
      .padding(5)
      .rotate(() => ~~(Math.random() * 2) * 90)
      .font('Arial')
      .spiral('rectangular')
      .fontSize((d: any) => d.size)
      .on('end', this.draw)
    this.layout.start()
  }
}
</script>
