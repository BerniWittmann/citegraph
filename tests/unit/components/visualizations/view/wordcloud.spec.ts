import { shallowMount } from '@vue/test-utils'
import { i18n } from '../../../setupPlugins'

import WordCloud from '@/components/visualizations/view/WordCloud.vue'
import WordCloudVisualization from '@/models/visualizations/WordCloudVisualization'

jest.useFakeTimers()

jest.mock('d3', () => {
  return {
    select: jest.fn().mockReturnThis(),
    append: jest.fn().mockReturnThis(),
    attr: jest.fn(function (attr, handler) {
      if (typeof handler === 'function') {
        handler({
          size: 12,
          x: 0,
          y: 19,
          rotate: 39,
          text: 'First'
        })
      }
      // @ts-ignore
      return this
    }),
    selectAll: jest.fn().mockReturnThis(),
    data: jest.fn().mockReturnThis(),
    enter: jest.fn().mockReturnThis(),
    style: jest.fn(function (attr, handler) {
      if (typeof handler === 'function') {
        handler({
          size: 12,
          x: 0,
          y: 19,
          rotate: 39,
          text: 'First'
        }, 1)
      }
      // @ts-ignore
      return this
    }),
    text: jest.fn().mockReturnThis(),
    scaleOrdinal: jest.fn().mockReturnValue(jest.fn())
  }
})
jest.mock('d3-cloud', () => {
  const result = {
    start: jest.fn(),
    size: jest.fn().mockReturnValue([500, 500])
  }
  const obj = {
    size: jest.fn().mockReturnThis(),
    words: jest.fn().mockReturnThis(),
    padding: jest.fn().mockReturnThis(),
    rotate: jest.fn(function (handler) {
      if (typeof handler === 'function') {
        handler()
      }
      // @ts-ignore
      return this
    }),
    font: jest.fn().mockReturnThis(),
    spiral: jest.fn().mockReturnThis(),
    fontSize: jest.fn(function (handler) {
      if (typeof handler === 'function') {
        handler({
          size: 12,
          x: 0,
          y: 19,
          rotate: 39,
          text: 'First'
        })
      }
      // @ts-ignore
      return this
    }),
    on: jest.fn(function (event, handler) {
      setTimeout(() => {
        handler([{
          size: 12,
          x: 0,
          y: 19,
          rotate: 39,
          text: 'First'
        }, {
          size: 32,
          x: 65,
          y: 65,
          rotate: 34,
          text: 'Second'
        }, {
          size: 23,
          x: 24,
          y: 64,
          rotate: 0,
          text: 'Thrid'
        }])
      }, 1)
      return result
    }),
    start: jest.fn().mockReturnThis()
  }
  return jest.fn().mockReturnValue(obj)
})

describe('components/visualizations/view/WordCloud.vue', () => {
  const getWrapper = (vis: WordCloudVisualization) => {
    return shallowMount(WordCloud, {
      i18n,
      mocks: {
        $t: (key: string) => key
      },
      propsData: {
        visualization: vis
      }
    })
  }

  const visualization = new WordCloudVisualization({
    id: '1',
    name: 'My Word Cloud',
    progress: 1,
    data: {
      data: [{
        word: 'Test',
        weight: 12
      }, {
        word: 'Eins',
        weight: 20
      }, {
        word: 'Zwei',
        weight: 22
      }, {
        word: 'Drei',
        weight: 23
      }]
    }
  })

  it('renders', () => {
    const wrapper = getWrapper(visualization)
    jest.runAllTimers()
    expect(wrapper.html()).toMatchSnapshot()
  })
})
