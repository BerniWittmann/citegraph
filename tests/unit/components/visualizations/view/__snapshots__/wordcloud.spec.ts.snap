// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/visualizations/view/WordCloud.vue renders 1`] = `<div></div>`;
